### PR TITLE
Improve error checking in the command encoder.

### LINKED ISSUE #1644: Fix error checking in encoder.c/decoder.c
#### Problem
The error checking in `encoder.c` and `decoder.c` checks that there is enough buffer to write or read (respectively) one more byte, but sometimes we are reading/writing multi-byte values (e.g. `uint16_t`).

 #### Proposed Solution
Fix the bounds-checking to check that there is enough buffer left for whatever size it is our data actually is.  It might make sense to factor out the "check the bounds, encode/decode, advance the cursor" bit into a macro or (templated?) function to avoid the duplication we have going on right now.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
There are actually two changes here:

1) Improve the error checking so that we correctly check that the
   buffer actually has enough space when writing a multibyte value.

2) Introduce a way to ask the encoder how much space it needs to
   encode a given APS frame.

Fixes https://github.com/project-chip/connectedhomeip/issues/1644

 #### Problem
The error checking in the command encoder doesn't work correctly when encoding multibyte values

 #### Summary of Changes
Fix the error checking to correctly check that we have enough space.

### CODE REVIEW

File: src/app/chip-zcl-zpro/command-encoder/encoder.c [hunk 1]
```diff
@@ -16,98 +16,67 @@
...
-    {
-        return size;
-    }
     size_t nextOutByte = 0;
 
-    if (nextOutByte >= buf_length)
-    {
-        return size;
-    }
+#define TRY_WRITE(dataItem)                                                                                                        \
```
Comment (BroderickCarlin) [inline src/app/chip-zcl-zpro/command-encoder/encoder.c:29]: Do we have a generic method for interfacing with buffers? Having proper methods for interfacing with buffers may be a better solution than a locally defined macro

Comment (rwalker-apple) [reply src/app/chip-zcl-zpro/command-encoder/encoder.c:29]: something that works on top of PacketBuffer would be welcome. something that has a "measure" form would be ideal.

Comment (rwalker-apple) [reply src/app/chip-zcl-zpro/command-encoder/encoder.c:30]: I've just added BufBound, which might help here

**Review by rwalker-apple — APPROVED** [review]

File: src/app/chip-zcl-zpro/command-encoder/encoder.c (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/chip-zcl-zpro/command-encoder/encoder.c:30]: We'd need to switch this file to C++ to use that.  Which might be fine as long as we keep the entrypoint APIs extern C....  I can rebase on top of your change once it lands as desired, and make that modification.

**Review by BroderickCarlin — APPROVED** [review]

**Review by bhaskar-apple — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@jelderton @andy31415 ?

**Review by jelderton — APPROVED** [review]

File: src/app/chip-zcl-zpro/command-encoder/encoder.c (hunk 1 above)
Comment (rwalker-apple) [reply src/app/chip-zcl-zpro/command-encoder/encoder.c:30]: ah, yeah.  I was thinking that we probably want to just make all these files C++ and fix any void* stuff

Comment (bzbarsky-apple) [reply src/app/chip-zcl-zpro/command-encoder/encoder.c:30]: Yeah, https://github.com/project-chip/connectedhomeip/issues/1629 tracks the `void*` bits.

Comment (bzbarsky-apple) [reply src/app/chip-zcl-zpro/command-encoder/encoder.c:30]: https://github.com/project-chip/connectedhomeip/issues/2073 now tracks using `BufBound` here.

**Pull request merged**